src/storage/src/region/writer.rs

```diff
@@ -220,6 +226,59 @@ impl RegionWriter {
 
         Ok(())
     }
+
+    pub async fn close<S: LogStore>(&self, writer_ctx: WriterContext<'_, S>) -> Result<()> {
+        // In order to close a writer
+
+        // 1. (Frontend) The frontend sets a CLOSING flag for the target region in metasrv. After that, it will reject future writing / altering requests.
+        // 2. (Frontend) Sends the close request to datanode leader(s).
+        // 2.1 (Datanode-leader) the leader acquires the write lock.
+        // 2.2 (Datanode-leader) Sets a memory flag to reject any potential(due to network delay) writing.
+        // 2.4 (Datanode-leader) Waits the pending flush task.
+        // 2.5 (Datanode-leader) Cancels compaction task (TODO).
+        // 2.6 (Datanode-leader) Notify the frontend.
+        // 3. (Frontend) The frontend sets a CLOSED flag for the target region in metasrv.
+
+        {
+            let mut inner = self.inner.lock().await;
+
+            if inner.is_closed() {
+                return Ok(());
+            }
+
+            inner.close();
+        }
+        // we release the writer lock once for rejecting any following potential writing requests immediately.
+
+        self.trigger_all_flush(writer_ctx).await?;
```

Comment:
While closing, we should cancel all background jobs to avoid blocking on flush or compaction for a long time.

Reply:
I got you. We could recover the pending writes from the WAL.